oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2TokenIntrospectionAuthenticationProvider.java

```diff
@@ -0,0 +1,178 @@
+ *
+ * @author Gerardo Roza
+ * @since 0.0.4
+ * @see OAuth2TokenIntrospectionAuthenticationToken
+ * @see OAuth2AuthorizationService
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7662#section-2.1">Section 2.1 - Introspection Request</a>
+ */
+public class OAuth2TokenIntrospectionAuthenticationProvider implements AuthenticationProvider {
+	private final OAuth2AuthorizationService authorizationService;
+	private final Collection<JwtDecoder> jwtDecoders;
+
+	/**
+	 * Constructs an {@code OAuth2TokenIntrospectionAuthenticationProvider} using the provided parameters.
+	 *
+	 * @param authorizationService the authorization service
+	 * @param jwtDecoders all available decoders that might be used to parse the token as a JWT-encoded token
+	 */
+	public OAuth2TokenIntrospectionAuthenticationProvider(OAuth2AuthorizationService authorizationService,
+			Collection<JwtDecoder> jwtDecoders) {
+		Assert.notNull(authorizationService, "authorizationService cannot be null");
+		this.authorizationService = authorizationService;
+		this.jwtDecoders = jwtDecoders;
+	}
+
+	@Override
+	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+		OAuth2TokenIntrospectionAuthenticationToken tokenIntrospectionAuthentication = (OAuth2TokenIntrospectionAuthenticationToken) authentication;
+
+		OAuth2ClientAuthenticationToken clientPrincipal = getAuthenticatedClientElseThrowInvalidClient(
+				tokenIntrospectionAuthentication);
+		RegisteredClient registeredClient = clientPrincipal.getRegisteredClient();
+
+		TokenType tokenType = null;
+
+		try {
+			String tokenTypeHint = tokenIntrospectionAuthentication.getTokenTypeHint();
+			if (StringUtils.hasText(tokenTypeHint)) {
+				if (TokenType.REFRESH_TOKEN.getValue().equals(tokenTypeHint)) {
+					tokenType = TokenType.REFRESH_TOKEN;
+				} else if (TokenType.ACCESS_TOKEN.getValue().equals(tokenTypeHint)) {
+					tokenType = TokenType.ACCESS_TOKEN;
+				} else if (TokenType.ID_TOKEN.getValue().equals(tokenTypeHint)) {
+					tokenType = TokenType.ID_TOKEN;
+				} else {
+					throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes2.UNSUPPORTED_TOKEN_TYPE));
+				}
+			}
+
+			OAuth2Authorization authorization = this.authorizationService
+					.findByTokenWithHint(tokenIntrospectionAuthentication.getTokenValue(), Optional.ofNullable(tokenType))
+					.orElseThrow(() -> new IntrospectionTokenException("Token not found"));
+
+			if (!registeredClient.getId().equals(authorization.getRegisteredClientId())) {
+				throw new IntrospectionTokenException("Invalid client");
+			}
+
+			AbstractOAuth2Token token = authorization.getTokens().getToken(tokenIntrospectionAuthentication.getTokenValue())
+					.get();
+
+			if (authorization.getTokens().getTokenMetadata(token).isInvalidated()) {
+				throw new IntrospectionTokenException("Token has been invalidated");
+			}
+
+			token = parseJwt(token);
```

Comment:
No need to `parseJwt()` or `validateToken()` as the original `Jwt` is stored in [OAuth2Authorization.attributes](https://github.com/spring-projects-experimental/spring-authorization-server/blob/8c71e5635089993be98997f3a21589b30f834056/oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2AuthorizationCodeAuthenticationProvider.java#L138)

Reply:
Two things regarding this @jgrandja :
1- We're currently storing the `access_token` `Jwt` in the OAuth2Authorization.attributes. However, we can also introspect a `refresh_token`, or in the future, an `id_token` as well.
What is the plan here? Are we going to store **all** `Jwt` token versions in OAuth2Authorization.attributes (which would explain why the `refresh_token` is not stored; it is not jwt-formatted), or just the `access_token`?
If that is, in fact, the plan, then we can certainly avoid parsing the `Jwt`, the logic, in this case, would be "we try to find the `Jwt` version in the `OAuth2Authorization.attributes` storage if we can't find it we validate it as a regular opaque token".
However, if we still don't have a clear plan, or if we want to gain flexibility (avoid tying this endpoint functionality to what we store in the `OAuth2Authorization.attributes` then the current approach might be more suitable; we simply try to decode the tokenValue using the same configuration we used to encode it.
2- I think we do need the `validateToken()` in any case, to run any validations we consider suitable at the time of introspecting the token.

Reply:
A quick update here regarding point 2 above.
I realized the `decode` operation we're using now also validates the `Jwt`, so there is no need to validate ourselves in that case.
However, if it is not a JWT-formatted token, then we still want to validate it (using the `AbstractOAuth2Token` fields) to make sure it is not expired for example.
BTW, the logic changed a little bit after we remove the CryptoKeySource (https://github.com/spring-projects-experimental/spring-authorization-server/issues/196) So have a look at the updated changes.

Reply:
There has been a lot of changes to the `OAuth2Authorization` model and it's associated tokens via #213 and subsequent commits.

The important thing to note here is that whatever is returned in `authorizationService.findByToken(tokenValue, null)` is the "source of truth" and the tokens contained within should be used on the code path forward.

@rozagerardo Before I continue with the review, please rebase off of master and integrate the changes. Also, I noticed you did a merge commit - please avoid this and only rebase. Thanks.

Reply:
Cool, I had a look at the new approach, I think this allows a cleaner approach for the Introspection Endpoint. I now adapted the code for these changes.